File: src/test/java/kitchenpos/eatinorders/tobe/ordertable/domain/OrderTableTest.java
package kitchenpos.eatinorders.tobe.ordertable.domain;

import kitchenpos.OrderTableFixture;
import org.assertj.core.api.ThrowableAssert;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.NullAndEmptySource;
import org.junit.jupiter.params.provider.ValueSource;

import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatExceptionOfType;

@DisplayName("주문 테이블(OrderTable)은")
class OrderTableTest {
    private static final UUID id = UUID.randomUUID();
    private static final String name = "주문테이블";

    @DisplayName("생성할 수 있다.")
    @Test
    void create() {
        final OrderTable orderTable = new OrderTable(id, name);
        assertThat(orderTable).isEqualTo(new OrderTable(id, name));
        assertThat(orderTable.isEmpty()).isTrue();
    }

    @DisplayName("비어있는 이름으로 생성시 IllegalArgumentException이 발생한다.")
    @ParameterizedTest
    @NullAndEmptySource
    void create(final String name) {
        ThrowableAssert.ThrowingCallable throwingCallable = () -> new OrderTable(name);
        assertThatExceptionOfType(IllegalArgumentException.class)
                .isThrownBy(throwingCallable);
    }

    @DisplayName("빈 테이블을 해지할 수 있다.")
    @Test
    void sit() {
        final OrderTable orderTable = OrderTableFixture.주문테이블();
        assertThat(orderTable.isEmpty()).isTrue();

        orderTable.sit();
        assertThat(orderTable.isEmpty()).isFalse();
    }

    @DisplayName("방문 손님수를 변경할 수 있다.")
    @Test
    void changeNumberOfGuests() {
        final OrderTable orderTable = OrderTableFixture.앉은테이블();
        assertThat(orderTable.isEmpty()).isFalse();

        orderTable.changeNumberOfGuests(3);
        assertThat(orderTable.getNumberOfGuests()).isEqualTo(3);
    }

    @DisplayName("변경하려는 손님수가 0명 미만이면 IllegalArgumentException이 발생한다.")
    @ParameterizedTest
    @ValueSource(strings = "-1")
    void changeNumberOfGuests(final int numberOfGuests) {
        final OrderTable orderTable = OrderTableFixture.앉은테이블();

        ThrowableAssert.ThrowingCallable throwingCallable = () -> orderTable.changeNumberOfGuests(numberOfGuests);
        assertThatExceptionOfType(IllegalArgumentException.class)
                .isThrownBy(throwingCallable);
    }

    @DisplayName("빈 테이블에 방문 손님수를 변경하려하면 IllegalStateException이 발생한다.")
    @Test
    void changeNumberOfGuests_empty() {
        final OrderTable orderTable = OrderTableFixture.주문테이블();
        assertThat(orderTable.isEmpty()).isTrue();

        ThrowableAssert.ThrowingCallable throwingCallable = () -> orderTable.changeNumberOfGuests(3);
        assertThatExceptionOfType(IllegalStateException.class)
                .isThrownBy(throwingCallable);
    }

    @DisplayName("빈 테이블로 설정할 수 있다.")
    @Test
    void clear() {
        final OrderTable orderTable = OrderTableFixture.앉은테이블();
        assertThat(orderTable.isEmpty()).isFalse();

        orderTable.clear();
        assertThat(orderTable.isEmpty()).isTrue();
    }
}

File: src/test/kotlin/kitchenpos/tobe/domain/eatinorder/domain/OrderTableTest.kt
package kitchenpos.tobe.domain.eatinorder.domain

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.shouldBe
import kitchenpos.tobe.eatinorder.domain.entity.OrderTableV2
import tobe.domain.OrderTableV2Fixtures.createOrderTableV2
import java.util.*

class OrderTableTest() : DescribeSpec() {
    init {
        describe("OrderTable 클래스의") {
            describe("from 메소드는") {
                it("OrderTable 객체를 생성한다.") {
                    val id = UUID.randomUUID()
                    val name = "테이블1"
                    val numberOfGuests = 4
                    val occupied = false

                    val orderTable = OrderTableV2.from(id, name, numberOfGuests, occupied)

                    orderTable.id shouldBe id
                    orderTable.name shouldBe name
                    orderTable.numberOfGuests shouldBe numberOfGuests
                    orderTable.occupied shouldBe occupied
                }
            }

            describe("clear 메소드는") {
                it("주문 테이블의 속성을 초기화한다.") {
                    val orderTable = createOrderTableV2()

                    orderTable.clear()

                    orderTable.numberOfGuests shouldBe 0
                    orderTable.occupied shouldBe false
                }
            }

            describe("sit 메소드는") {
                it("주문 테이블을 점유한다.") {
                    val orderTable = createOrderTableV2()

                    orderTable.sit()

                    orderTable.occupied shouldBe true
                }
            }

            describe("changeNumberOfGuests 메소드는") {
                context("적절한 손님 수가 주어지면") {
                    it("주문 테이블의 손님 수를 변경한다.") {
                        val orderTable =
                            createOrderTableV2().also {
                                it.occupied = true
                            }
                        val numberOfGuests = 5

                        orderTable.changeNumberOfGuests(numberOfGuests)

                        orderTable.numberOfGuests shouldBe numberOfGuests
                    }
                }

                context("0명 미만의 손님 수가 주어지면") {
                    it("예외를 발생시킨다.") {
                        val orderTable = createOrderTableV2()
                        val numberOfGuests = -1

                        val exception =
                            shouldThrow<IllegalArgumentException> {
                                orderTable.changeNumberOfGuests(numberOfGuests)
                            }

                        exception.message shouldBe "고객 수는 0명 이상이어야 합니다."
                    }
                }

                context("주문 테이블이 점유되지않은 경우") {
                    it("예외를 발생시킨다.") {
                        val orderTable = createOrderTableV2()
                        val numberOfGuests = 5

                        val exception =
                            shouldThrow<IllegalArgumentException> {
                                orderTable.changeNumberOfGuests(numberOfGuests)
                            }

                        exception.message shouldBe "점유되지않은 주문 테이블의 고객 수는 변경할 수 없습니다."
                    }
                }
            }
        }
    }
}
